### PR TITLE
`Logos`: UI Enhancements and Fair lane Eviction Algo

### PR DESCRIPTION
**Pull request by JakubJakubczyk**
# Logos UI Integration, ETTFT-Corrected Classification Scheduling, and Capacity Planner Hardening

## Summary

This PR brings three major workstreams together:

1. **Logos UI remaster** — a rebuilt `/statistics` page with per-worker GPU panels, live lane-health metrics, VRAM allocation pie, and paginated request history wired to the new v2 stats WebSocket.
2. **ETTFT-Corrected Classification Scheduling (ECCS)** — a reworked ETTFT estimator with a range-scaled additive penalty formulation, VRAM-aware reclaim tiers, multi-provider candidate expansion, and decision logging for ablation studies.
3. **Capacity planner hardening** — replaces forward VRAM reservations with a per-provider capacity lock, adds anti-thrashing with tenure + adaptive cooldown + queue-depth fairness, VRAM headroom gate, sleeping-lane reclaim, multi-lane drain, and UNAVAILABLE fallback

Also included: dynamic logosnode deployment migration, worker-node p50 latency plumbing into the SDI, large benchmark/ablation infrastructure, and new architecture docs.

---

## 1. UI integration (logos-ui)

New `/statistics` page replaces the previous layout with a component-driven architecture.

### New components (`logos-ui/components/statistics/`)
- `worker-gpu-panel.tsx` — per-worker GPU card with connection state, device info, VRAM bars, and loaded/sleeping/cold lane counts.
- `lane-metrics-panel.tsx` — live lane-health rows with state dot, runtime mode (vLLM badge), KV-cache usage bar (green/amber/red thresholds), and p95 TTFT color coding.
- `lane-vram-pie.tsx` / `lane-vram-pie.web.tsx` — Plotly VRAM allocation pie (web) with native fallback.
- `paginated-request-list.tsx` — paginated request history with filtering.
- `types.ts` — shared types for `DeviceInfo`, `LaneSignalData`, and related scheduler-signal payloads.
- `constants.ts` — centralized lane-state color palette and API base.

### `app/statistics.tsx` (rewrite)
- 40 / 60 column split, combined GPU + VRAM card, lane-health section, compact summary-header cards.
- Dark-mode theming fixes across the pie chart and new panels.
- New skeletons for all new sections.

### `hooks/use-stats-websocket-v2.ts`
- Typed `VramV2Sample` / `VramV2Provider` payloads with `scheduler_signals.provider`, `scheduler_signals.lanes`, and `devices[]`.
- Tracks connection state, runtime modes, and last heartbeat per provider.

### Backend stats fix
- `fix(stats)`: handle the renamed vLLM `kv_cache` metric (`f8e3227f`) so GPU cache usage continues to render after vLLM's metric rename.

---

### `ettft_estimator.py` — full rewrite
Replaces the fixed-penalty tier table with a **range-scaled additive correction**:

```
corrected(m, p) = w(m) - penalty(m, p)
penalty        = min(E[wait] / H, 1.0) × S × α
```

- `H` = 60 s normalization horizon, `α` = 1.5 correction strength.
- `S` = weight span with a 20 % floor and absolute floor of 1.0, so close or negative weights still yield meaningful corrections.
- `E[wait] = state_overhead + reclaim_overhead + queue_wait`, where queue wait uses observed per-request service time and the 3× vLLM effective-parallel multiplier.
- Six VRAM-aware tiers: `WARM`, `SLEEPING`, `SLEEPING_RECLAIM`, `COLD`, `COLD_RECLAIM`, `UNAVAILABLE`. Reclaim overhead distinguishes idle eviction (3 s) from busy drain (~30 s).
- Preserves same-state classification ordering (same state → same penalty).
- Backward-compatible `ettft_ms` property for older callers.

### `correcting_scheduler.py`
- All schedulers (FCFS / utilization / correcting) now enumerate **every `(model_id, provider_id)` deployment** instead of `next()`-matching the first. Each pair is an independent scored candidate.
- Queue fallback prefers logosnode deployments.
- Lane-readiness gating: skips candidates whose lanes are not in a ready state.
- **ECCS decision logging** — JSON-lines output (env `ECCS_DECISION_LOG`) with candidates, classification weights, corrected scores, tiers, selection, and a `correction_changed` flag for ablation.
- **Weight override** — env `ECCS_WEIGHT_OVERRIDE` injects known weights for controlled experiments.

### Supporting changes
- `sdi/logosnode_facade.py`: new `get_parallel_capacity`, `get_model_name`, `get_scheduler_queue_depth_by_model_name`, `is_model_lane_ready`.
- `sdi/models.py`, `sdi/providers/logosnode_provider.py`, `logos/logosnode_registry.py`, `logos_worker_node/vllm_process.py`: **p50 latency** propagated end-to-end into `LaneSchedulerSignals` and into `ettft_estimator` as observed service time.
- `capacity/capacity_planner.py`: lane-ready signal fed to ETTFT.
---

## 3. Capacity planner — big fixes

Chronological fixes to `src/logos/capacity/capacity_planner.py` (1 562 lines changed net):

| Commit | Issue fixed | Result |
|---|---|---|
| `cf4c51c0` | Thrashing under dual-model GPU contention | ~99 → ~6 drains; min-tenure, adaptive cooldown, queue fairness, fire-and-forget triggers, lane-readiness gating |
| `f57d2747` | `release()` TOCTOU + insufficient resolver retries | Lane tenure protection added; context resolver retry window extended |
| `9febf08b` | Forward VRAM reservation deadlocks (committed VRAM ballooned to 48 GB on a 32 GB GPU) | Replaced with per-provider `asyncio.Lock`; 147/150 success, 0 cold starts, 0 lock timeouts |
| `74852e75` | Demand-blind reclaim, per-lane VRAM gate blocking multi-lane drain, vestigial thrash code, mismatched drain-reason string | Demand-gated reclaim, multi-lane drain, dead-code cleanup; 148/150 success, 0 thrash incidents |
| `bcf78771` | vLLM under-utilized (worst-case-context concurrency) | `VLLM_CONCURRENCY_OVERSUBSCRIPTION=3` (Qwen 7B 36→108, Qwen 14B 10→30, Mistral 7B 16→48); `BACKEND_QUEUE_PRESSURE_THRESHOLD` 2→8; `_drain_lane` also checks vLLM `queue_waiting` / `requests_running`. **P50 90.3 s → 61.0 s (−32 %)**, P95 −11 %, 0 % failures |
| `11ea3774` | Drain overshoot during continuous batching | Context resolver retry 30 s → 60 s. 100 % success at 150/300/600 vs 98.7 % / 81 % / 76 % |
| `cc38dcd0` | Unfair model switching when scheduler queue is skewed; no fallback when provider becomes `UNAVAILABLE` | Queue-fair model switching + UNAVAILABLE fallback + anti-thrash v4 |
| `4431461a` | vLLM utilization floor causing early evictions; sleeping lanes not being reclaimed | VRAM headroom gate; sleeping-lane reclaim; vLLM utilization floor disabled; multi-provider scheduling |
| `da59856b` | Swap fairness | Stable fairness algo for swaps, secure VRAM allocations/loads |

### Supporting
- `capacity/vram_ledger.py`: cleaned up reservations API.
- `logosnode_registry.py`, `main.py`: dynamic worker announcements, capacity trigger retry.
- `monitoring/prometheus_metrics.py`: new counters (switch events, reclaim reasons).

---

## 4. Worker node

- `lane_manager.py`, `logos_bridge.py`, `calibration.py`, `model_profiles.py`, `vllm_process.py`: p50 latency, per-GPU KV-cache size calculation fix (`5b3de664`), dynamic model announcement support.
- `config.yml` / `model_profiles.yml`: updated tuning (3× oversubscription, queue threshold 8, lane profiles).
- `docker-compose.dev.yml`: `sdi/` bind-mount for live code reload.

---

## 5. Database

- `db/migrations/027_logosnode_dynamic_deployments.sql` — adds `logosnode_provider_keys` table so workernode providers don't need per-model rows in `model_api_keys`; application layer auto-syncs `model_provider` rows from WebSocket capabilities.
- `db/init.sql` updated with the new table so fresh dev environments match production.

---

## 6. Infrastructure / benchmarks / ablation

- `tests/performance/run_eccs_benchmark.py`, `run_eccs_ablation.sh`, `analyze_eccs_ablation.py`, `generate_eccs_ablation_workloads.py`, `plot_benchmark_distributions.py`, `plot_ollama_vs_vllm{,_3x,_ttft}.py`, `run_ollama_benchmark.py`, `run_ollama_benchmark_suite.sh`, `benchmark_spread.sh`, `regenerate_charts.py`.
- New workloads under `performance/workloads/ablation_5model/`, `explicit/10m/`, `resource/10m/`.
- New `comparative/` Ollama-vs-vLLM result sets (150/300/600) and refreshed `logosworkernode-singular/` results.
- `tests/integration/test_capacity_fixes.sh`, `tests/smoke_test_capacity.sh`.
- Cleaned up the previous `explicit/` result folders (replaced by the organized 6-folder resolver30s/60s layout).

### Unit tests updated
- `tests/unit/capacity/test_capacity_planner.py` (+335 lines)
- `tests/unit/pipeline/test_correcting_scheduler.py` (+565 lines)
- `tests/unit/pipeline/test_ettft_estimator.py` (+517 lines)

---

## 7. Docs

- `logos/docs/capacity-management-architecture.md` (1 544 lines) — end-to-end capacity planner architecture.
- `logos/docs/ettft-scheduling.md` (205 lines) — ECCS formal model.
- `logos/docs/benchmark-results-analysis.md` (169 lines) — benchmark methodology & results.
---

### CODE REVIEW

File: logos/src/logos/queue/priority_queue.py [hunk 1]
```diff
@@ -3,9 +3,6 @@
 
 This module provides thread-safe priority queue operations for scheduling.
 Maintains separate heaps per model per priority level.
-
-Queue key: model_id only (provider_id was dropped so queued requests can
```
Comment (wasnertobias) [inline logos/src/logos/queue/priority_queue.py:7]: This is still relevant "provider_id was dropped so queued requests can re-route across workers as they become ready"

File: logos/src/logos/sdi/providers/logosnode_provider.py [hunk 2]
```diff
@@ -31,7 +33,7 @@
 class LogosNodeDataProvider:
     """Unified local-provider SDI source for direct Ollama and worker-backed lanes."""
 
-    DEFAULT_PARALLEL_CAPACITY = 20
+    DEFAULT_PARALLEL_CAPACITY = 1
```
Comment (wasnertobias) [inline logos/src/logos/sdi/providers/logosnode_provider.py:36]: This is way too less!

File: .github/workflows/logos_build-and-push-docker.yml [hunk 3]
```diff
@@ -96,20 +96,42 @@ jobs:
       docker-context: ./logos/db
       tags: ${{ needs.compute-tag.outputs.is_release == 'true' && format('ghcr.io/ls1intum/edutelligence/logos-db:latest,ghcr.io/ls1intum/edutelligence/logos-db:{0}', needs.compute-tag.outputs.release_tag) || format('ghcr.io/ls1intum/edutelligence/logos-db:{0}', needs.compute-tag.outputs.result) }}
 
+  # Build the vLLM workernode for amd64 only. vLLM is x86_64/CUDA-only at
```
Comment (wasnertobias) [inline .github/workflows/logos_build-and-push-docker.yml]: "Build the vLLM workernode for amd64 only" => sadly this change is incompatible with our existing CI/CD workflow

**Review by wasnertobias — CHANGES_REQUESTED** [review]

File: logos/src/logos/sdi/providers/logosnode_provider.py (hunk 2 above)
Comment (JakubJakubczyk) [reply logos/src/logos/sdi/providers/logosnode_provider.py:36]: This is effectively a fallback value for a brief moment when lane has not yet been created. In production, as long as you allocate enough kv-cache and for a given model a paralellity is set in the database, this value is not enforcing anything.

File: logos/src/logos/queue/priority_queue.py (hunk 1 above)
Comment (JakubJakubczyk) [reply logos/src/logos/queue/priority_queue.py:7]: This isn't quite as simple as dropping a key, I think. logos-server doesn't
  just forward requests; it also drives worker state. A model can't become
  warm and thus dispatch-eligible without logos-server issuing a load/wake                                                                                           
  to a specific provider. Decoupling the queue from provider_id therefore
  opens a new error surface:                                                                                                                          
                                           
    - avoiding provider starvation,                                                                                                                                  
    - preventing a single demand signal from triggering duplicate loads on
      multiple eligible providers,                                                                                                                                   
    - deciding when to replicate model A across providers for load-balancing
      vs. reusing whichever provider already has it warm to avoid cold-starts                                                                                        
      and have less forced evictions through more conservative memory
      management.                                                                                              
                                                                                                                                                                     
  Today the planner already does some of this multi-provider arbitration:                                                                                            
  when one provider would need an eviction to load, the work falls to                                                                                                
  another eligible provider whose score is better. Replicating that logic
  on top of a model_id-only queue is non-trivial, and I can't realistically                                                                                          
  verify the edge cases on the single-vLLM setup we have for testing.

**Review by wasnertobias — DISMISSED** [review]
Tested on test server, lgtm!
